core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java

```diff
@@ -19,115 +19,328 @@
-        this.distance = distance;
+        if (Strings.isEmpty(distance)) {
+            throw new IllegalArgumentException("distance must not be null or empty");
+        }
+        return this.distance(distance, DistanceUnit.DEFAULT);
+    }
+
+    /** Sets the distance from the center for this query. */
+    public GeoDistanceQueryBuilder distance(String distance, DistanceUnit unit) {
+        if (Strings.isEmpty(distance)) {
+            throw new IllegalArgumentException("distance must not be null or empty");
+        }
+        if (unit == null) {
+            throw new IllegalArgumentException("distance unit must not be null");
+        }
+        this.distance = DistanceUnit.parse(distance, unit, DistanceUnit.DEFAULT);
         return this;
     }
 
+    /** Sets the distance from the center for this query. */
     public GeoDistanceQueryBuilder distance(double distance, DistanceUnit unit) {
-        this.distance = unit.toString(distance);
+        if (unit == null) {
+            throw new IllegalArgumentException("distance unit must not be null");
+        }
+        this.distance = DistanceUnit.DEFAULT.convert(distance, unit);
         return this;
     }
 
+    /** Returns the distance configured as radius. */
+    public double distance() {
+        return distance;
+    }
+
+    /** Sets the center point for this query. */
     public GeoDistanceQueryBuilder geohash(String geohash) {
-        this.geohash = geohash;
+        if (Strings.isEmpty(geohash)) {
+            throw new IllegalArgumentException("geohash must not be null or empty");
+        }
+        this.center.resetFromGeoHash(geohash);
         return this;
     }
 
+    /** Which type of geo distance calculation method to use. */
     public GeoDistanceQueryBuilder geoDistance(GeoDistance geoDistance) {
+        if (geoDistance == null) {
+            throw new IllegalArgumentException("geoDistance must not be null");
+        }
         this.geoDistance = geoDistance;
         return this;
     }
 
+    /** Returns geo distance calculation type to use. */
+    public GeoDistance geoDistance() {
+        return this.geoDistance;
+    }
+
+    /**
+     * Set this to memory or indexed if before running the distance
+     * calculation you want to limit the candidates to hits in the
+     * enclosing bounding box.
+     **/
     public GeoDistanceQueryBuilder optimizeBbox(String optimizeBbox) {
-        this.optimizeBbox = optimizeBbox;
+        if (Strings.isEmpty(optimizeBbox)) {
+            throw new IllegalArgumentException("optimizeBbox must not be null or empty");
+        }
+        this.optimizeBbox = (optimizeBbox != null) ? optimizeBbox : DEFAULT_OPTIMIZE_BBOX;
         return this;
     }
 
+    /**
+     * Returns whether or not to run a BoundingBox query prior to
+     * distance query for optimization purposes.*/
+    public String optimizeBbox() {
+        return this.optimizeBbox;
+    }
+
     public GeoDistanceQueryBuilder coerce(boolean coerce) {
```

Comment:
`coerce` and `ignoreMalformed` seems to be related maybe we can enforce that both are updated together? or at least add a TODO